float32 x_error
float32 y_error
float32 distance_error